libcontainer/cgroups/file.go

```diff
@@ -137,6 +138,17 @@ func openFile(dir, file string, flags int) (*os.File, error) {
 			Mode:    uint64(mode),
 		})
 	if err != nil {
+		// Check if cgroupFd is still opened to cgroupfsDir.
+		fdStr := strconv.Itoa(cgroupFd)
+		fdDest, _ := os.Readlink("/proc/self/fd/" + fdStr)
+		if fdDest != cgroupfsDir {
+			// TODO: reopen cgroupFd and retry openat2.
+
+			// Enhance the Path in the error to contain the
+			// cgroupFd value and the directory it is opened to,
+			// for example: "@[fd 7:/!=/sys/fs/cgroup]/cpu.stat".
```

Comment:
Is this convention already widely used, or newly invented here?

Rather than defining a new microformat, can we just wrap `os.PathError` with our own struct?

Reply:
> Is this convention already widely used, or newly invented here?

Newly invented (originally comes from https://github.com/opencontainers/runc/issues/3026#issuecomment-860148103)

We are already using "fd N" in a few places where we operate on a file descriptor:
```console
[kir@kir-rhat runc]$ git grep PathError | grep -w fd
libcontainer/rootfs_linux.go:		return &os.PathError{Op: "fchdir", Path: "fd " + strconv.Itoa(newroot), Err: err}
libcontainer/rootfs_linux.go:		return &os.PathError{Op: "fchdir", Path: "fd " + strconv.Itoa(oldroot), Err: err}
libcontainer/setns_init_linux.go:		return &os.PathError{Op: "close log pipe", Path: "fd " + strconv.Itoa(l.logFd), Err: err}
libcontainer/standard_init_linux.go:		return &os.PathError{Op: "close log pipe", Path: "fd " + strconv.Itoa(l.logFd), Err: err}
vendor/github.com/opencontainers/selinux/go-selinux/selinux_linux.go:		return "", &os.PathError{Op: "getsockopt", Path: "fd " + strconv.Itoa(int(fd)), Err: err}
```

but in this case something more useful is needed.

> Rather than defining a new microformat, can we just wrap os.PathError with our own struct?

Done

Reply:
@AkihiroSuda PTAL ^^^